.github/workflows/pull_request.yml

```diff
@@ -36,14 +36,13 @@ jobs:
       - {uses: actions/checkout@v2, with: {fetch-depth: 0}}
       - uses: actions/setup-python@v2
 
-      # Package index is "out of date by design"
-      # https://github.com/actions/virtual-environments/issues/1757
-      - name: apt-get -y update
-        run: sudo apt-get -y update
+      # Use pip to install packages because apt seems to install
+      # the packages for Python whose path is different than setup-python use.
```

Comment:
Please be more specific, what exact error do you get with `apt` packages?

Reply:
<https://github.com/thesofproject/sof-test/blob/40b7f17a42be55d3480e209b13d5a3577ab34a4d/.github/workflows/pull_request.yml#L41-L42>
Is this OK?

Reply:
"Python import error" is much more specific than "path is different", now I understand thank you.

Reply:
Timely: https://lwn.net/Articles/867657/ Cooperative package management for Python

Reply:
I haven't subscribed to LWN.net, so I don't know what the article said. But I think maybe my words still mislead you. The actual problem is: `apt` always installs packages for Python in the system path (typically `/usr/bin`), while [setup-python](https://github.com/marketplace/actions/setup-python) setup Python in the other location (for us, this is `/opt/hostedtoolcache/Python/3.9.6/x64`, as you can see in the logs), so the packages installed by `apt` can't be found by Python we use. But `pip`  is distributed with Python, so `pip` we use is for Python we use.

Reply:
"We" don't install anything  `/opt/hostedtoolcache`, github does. That's not supposed to be a visible change. As the previous comment says, in the past it was only a problem for `apt-get install python3-numpy; import numpy` but it was not a problem for the others like `apt-get install python3-scipy; import scipy`. So this looked like a numpy-specific bug, either from Ubuntu or from github.

As this is a workaround for that bug, it would be great to give the exact list of the failing imports. Or even better: report this issue to Ubuntu or github and link to the bug.

Reply:
I tried to perform a simple test: <https://github.com/thesofproject/sof-test/pull/755/checks?check_run_id=3481619127>.
As you can see in the log, `import scipy` also failed with `apt-get`.

As my previous comments said, **all** the packages installed by `apt` can't be imported by Python located in `/opt/hostedtoolcache`.

Reply:
I just tested that again and now both numpi and scipy imported just fine:

https://github.com/thesofproject/sof-test/pull/759/checks?check_run_id=3487197768

Maybe github fixed these issues.

Reply:
Well, I find the difference between our tests: in your test, everything is installed via `apt`, including pylint, so it works fine; in my test, the packages are installed via `apt`, but use Python located in `/opt/hostedtoolcache`, so the packages can't be found. If we just use `apt`, we even don't need [setup-python](https://github.com/marketplace/actions/setup-python), just keep the Python preinstalled in Ubuntu.

Reply:
Thanks for solving this mystery! I didn't pay attention to https://github.com/actions/setup-python, my bad. It looks powerful and flexible but more complicated, simpler if we can avoid it.